
<!DOCTYPE HTML>
<html lang="" >
    <head>
        <meta charset="UTF-8">
        <meta content="text/html; charset=utf-8" http-equiv="Content-Type">
        <title>Command Line Manual Â· GitBook</title>
        <meta http-equiv="X-UA-Compatible" content="IE=edge" />
        <meta name="description" content="">
        <meta name="generator" content="GitBook 3.2.2">
        
        
        
    
    <link rel="stylesheet" href="../gitbook/style.css">

    
            
                
                <link rel="stylesheet" href="../gitbook/gitbook-plugin-highlight/website.css">
                
            
                
                <link rel="stylesheet" href="../gitbook/gitbook-plugin-search/search.css">
                
            
                
                <link rel="stylesheet" href="../gitbook/gitbook-plugin-fontsettings/website.css">
                
            
        

    

    
        
    
        
    
        
    
        
    
        
    
        
    

        
    
    
    <meta name="HandheldFriendly" content="true"/>
    <meta name="viewport" content="width=device-width, initial-scale=1, user-scalable=no">
    <meta name="apple-mobile-web-app-capable" content="yes">
    <meta name="apple-mobile-web-app-status-bar-style" content="black">
    <link rel="apple-touch-icon-precomposed" sizes="152x152" href="../gitbook/images/apple-touch-icon-precomposed-152.png">
    <link rel="shortcut icon" href="../gitbook/images/favicon.ico" type="image/x-icon">

    
    <link rel="next" href="tutorial.html" />
    
    
    <link rel="prev" href="GUI.html" />
    

    </head>
    <body>
        
<div class="book">
    <div class="book-summary">
        
            
<div id="book-search-input" role="search">
    <input type="text" placeholder="Type to search" />
</div>

            
                <nav role="navigation">
                


<ul class="summary">
    
    

    

    
        
        
    
        <li class="chapter " data-level="1.1" data-path="../">
            
                <a href="../">
            
                    
                    Preface
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.2" data-path="install.html">
            
                <a href="install.html">
            
                    
                    installation
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.3" data-path="intro.html">
            
                <a href="intro.html">
            
                    
                    introduction
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.4" data-path="GUI.html">
            
                <a href="GUI.html">
            
                    
                    GUI manual
            
                </a>
            

            
        </li>
    
        <li class="chapter active" data-level="1.5" data-path="manual.html">
            
                <a href="manual.html">
            
                    
                    Command Line Manual
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.6" data-path="tutorial.html">
            
                <a href="tutorial.html">
            
                    
                    Tutorial
            
                </a>
            

            
        </li>
    

    

    <li class="divider"></li>

    <li>
        <a href="https://www.gitbook.com" target="blank" class="gitbook-link">
            Published with GitBook
        </a>
    </li>
</ul>


                </nav>
            
        
    </div>

    <div class="book-body">
        
            <div class="body-inner">
                
                    

<div class="book-header" role="navigation">
    

    <!-- Title -->
    <h1>
        <i class="fa fa-circle-o-notch fa-spin"></i>
        <a href=".." >Command Line Manual</a>
    </h1>
</div>




                    <div class="page-wrapper" tabindex="-1" role="main">
                        <div class="page-inner">
                            
<div id="book-search-results">
    <div class="search-noresults">
    
                                <section class="normal markdown-section">
                                
                                <h3 id="main-script---obtain-principal-components-from-a-series-of-2d-measurements">Main Script - Obtain Principal Components from a series of 2D Measurements</h3>
<ul>
<li><h4 id="both-gui-graphic-user-interface-and-cli-command-line-interface-programs">Both GUI (Graphic User Interface) and CLI (Command Line Interface) programs</h4>
<p>GUI: <code>trendmaingui.exe</code> in Windows and Linux, <code>trendmaingui.app</code> in OS X</p>
<p>CLI: <code>trendmain.exe</code> on all platforms</p>
</li>
<li><h4 id="synopsis">Synopsis</h4>
<p><code>trendmain.exe -t [ucsf/fid/ft2/png/movie/txt/sparky/csv/excel/&#x2026;] -f [file index]
-r [threshold] -o [output prefix name] -s [scaling methods] [-x [xaxis file]] [-u [Unit]]
[--report]</code></p>
</li>
<li><h4 id="description">Description</h4>
<p><code>trendmain.exe</code> reads a series of NMR spectra, FIDs, png images, the 
Joint Committee on Atomic and Molecular Physical data - Data eXchange 
(JCAMP-DX) format, plain text files, comma-separated (CSV) or Excel 
spreadsheets listed in the index file.</p>
<p><code>trendmain.exe</code> also reads and process single file, including movie, CSV or 
Excel spreadsheet, or text file.  TREND preprocesses them, applies singular 
value decomposition to provide U, s, and V matrices (Xu and Van Doren, 
submitted). Its normal mode is to provide principal components (PCs) from the 
first few rows of the V matrix. TREND also has options to perform PCA 
reconstructions of the original data series or to perform independent components
analysis. s, V, and normalized PC1 values are output as text files in the names
of <code>prefix-s.txt</code>, <code>prefix-vt.txt</code>, and <code>prefix-pc1.txt</code>, which reports the binding 
isotherm. Additional PCs can be requested.</p>
</li>
<li><h4 id="typical-usage-with-nmr-spectra">Typical usage with NMR spectra:</h4>
<p><code>trendmain.exe -t ucsf -f file.index -r auto -o outfile -s auto --report</code></p>
</li>
<li><h4 id="options">Options</h4>
<ul>
<li><p><code>-t [ucsf/fid/ft2/png/movie/txt/complextxt/csv/complexcsv/singlematrix]</code></p>
<p><code>-t</code> specifies the input file format, the default format is .ucsf for NMR 
spectra. The options for input file formats are:</p>
<ul>
<li><strong>ucsf</strong> Sparky ucsf format (default)  </li>
<li><strong>fid</strong> NMRPipe fid format  </li>
<li><strong>ft2</strong> NMRPipe ft2 format  </li>
<li><strong>png</strong> image format (needs Scipy module)  </li>
<li><strong>brukerfid</strong> Bruker Topspin FID format (<code>fid</code> or <code>ser</code> file 
in the experimental directory) </li>
<li><strong>brukerspectra</strong>  Bruker Topspin processed spectra (<code>1r</code> (1D) or 
<code>2rr</code> (2D) in the <code>1/pdata/</code> subdirectory in the experimental directory . 
Note currently the processed spectra must be saved by setting 
processing number as 1 in the Topspin experimental directory    </li>
<li><strong>agilentfid</strong> Agilent VnmrJ (OpenVnmrJ) fid. (<code>fid</code> file in the 
data directory)  </li>
<li><strong>agilentspectra</strong> Agilent (Varian) VnmrJ (OpenVnmrj) fid. (<code>phasefile</code> 
in the <code>datdir</code> subdirectory. Note in order to make <code>phasefile</code> 
readable by third-party software including TREND, set trace=&apos;f1&apos;, 
display the full spectrum, and use the VnmrJ <code>flush</code> command are 
required. Otherwise the <code>phasefile</code> will be all zero values. See 
the <a href="https://www.cgl.ucsf.edu/home/sparky/manual/files.html" target="_blank">Sparky manual</a> 
for details.<br><strong>Note</strong> TREND supports on-the-fly analysis on Bruker and 
Agilent spectrometers. However to avoid effect of non-uniformity 
or outliers, all experimetnal data analyzed by TREND are preferably 
to be collected and processed in identical conditions. Therefore, 
we highly recommend using the same set of <code>NMRPipe</code> scripts to 
convert Bruker or Agilent data into NMRPipe <strong>fid</strong> and <strong>ft2</strong> formats.  </li>
<li><strong>txt</strong> Data stored in plain text format delimited by spaces, this 
option could be used for other spectroscopy methods such as time-
dependent 2D-IR.  </li>
<li><strong>csv</strong>  csv format is supported by Excel, OpenOffice and many 
software for spectroscopy instruments.  </li>
<li><strong>complextxt</strong> Similar to plain text format (<code>txt</code>) but using complex
number in the form of <code>0.00+0.00j</code>.  </li>
<li><strong>complexcsv</strong>  Similar to csv format. However, each cell should be in
the format of <code>0.00+0.00j</code>.  </li>
<li><strong>excel</strong> Microsoft Excel format (xlsx is preferred, the older xls
format is also accepted). All cells must be float or complex numbers. 
The complex number should be in the form of <code>0.00+0.00j</code>. By default an
Excel file contains three sheets, only the first sheet is analyzed. 
Excel with multi-sheets can be processed as <code>singleexcel</code> option.    </li>
<li><strong>sparky</strong> or <strong>sparkylist</strong> Sparky list format. The 1st column lists any residue 
assignments. The 2nd and 3rd columns are F1 (often 15N) and F2 (often 
1H) chemical shifts, respectively.  </li>
<li><strong>singlematrix</strong> This mode serves as a general PCA tool to process a 
single matrix in the text format delimited with spaces.   </li>
<li><strong>singlecsv</strong>  Similar to <code>singlematrix</code> format but in <code>csv</code> format</li>
<li><strong>singleexcel</strong>  Similar to <code>singlematrix</code> and <code>singlecsv</code> formats but
in the <code>xlsx</code> or <code>xls</code> formats. Besides, a series of 1D or 2D dataset 
can be stored in a single Excel file containing different, 
sequentially-ordered sheets.  </li>
<li><strong>movie</strong> Common video formats, such as <code>.ogv</code>, <code>.mp4</code>, <code>.mpeg</code>, 
<code>.avi</code>, <code>.mov</code>, <code>.webm</code>   </li>
<li><strong>jcamp</strong> JCAMP-DX is a general format for exchanging and archiving 
data from many instruments, including but not limited to IR, Raman, 
Uv-Vis, Fluorescence, NMR, and EMR. The data stored in JCAMP-DX 
files can be spectral plots, contours, or peak tables.  JCAMP-DX 
is very flexible in order to support most kinds of computerized 
instruments. As a result, it is impractical to support all 
existing JCAMP-DX variants. However, TREND supports most common 
JCAMP-DX formats. The digital data in JCAMP-DX can be <strong>AFFN</strong> 
(ASCII FREE FORMAT NUMERIC) form or <strong>ASDF</strong> (ASCII SQUEEZED 
DIFFERENCE FORM). TREND supports decoding compressed data, 
including <code>PAC</code>, <code>SQZ</code>, <code>DIF</code>, <code>SQZDUP</code>, and <code>DIFDUP</code>. Two most 
common tabular data form, <code>(X++(Y..Y))</code> and <code>(XY..XY)</code> are supported. 
TREND reads a series of JCAMP-DX files, or a single JCAMP-DX file 
with one or multiple blocks. TREND supports <code>NTUPLE</code> format 
(introduced by JCAMP-DX 5.0 standard), which is designed for 
multi-dimensional techniques with data sets of multiple variable. 
For example, JCAMP-DX NMR uses <code>NUTPLE</code> to show mixed real/imaginary 
FID data sets. See <a href="http://www.jcamp-dx.org/" target="_blank">http://www.jcamp-dx.org/</a> , 
<a href="https://badc.nerc.ac.uk/help/formats/jcamp_dx/" target="_blank">https://badc.nerc.ac.uk/help/formats/jcamp_dx/</a>
and <a href="http://wwwchem.uwimona.edu.jm:1104/spectra/testdata/index.html" target="_blank">http://wwwchem.uwimona.edu.jm:1104/spectra/testdata/index.html</a> 
for details of JCAMP-DX formats.   </li>
</ul>
</li>
<li><p><code>-f [fileindex/filename]</code></p>
<p><code>-f</code> reads file index and is required</p>
<ul>
<li><code>[fileindex]</code> is a text format index referring to the  series of NMR 
spectra, images, lists, or directories (for Bruker Topspin and 
Agilent VnmrJ data). The default file index name is <code>file.index</code>.
An example is a series of five .ucsf files named numerically. The index
file can be named as <code>index.ucsf</code> and includes the following lines in 
the example:<br>![file.index.png]
(<a href="https://bitbucket.org/repo/bboE8M/images/1070799385-file.index.png" target="_blank">https://bitbucket.org/repo/bboE8M/images/1070799385-file.index.png</a>)<br>Another example is a series of Bruker Topspin directories on a 
Windows platform. Absolute paths are supported as well as relative 
path, which is shown in the <code>UCSF</code> example. This &quot;directory list&quot; 
can be used for both <strong>brukerfid</strong> and <strong>brukerspectra</strong>. For 
<strong>brukerfid</strong> data, TREND will read <code>fid</code> or <code>ser</code> data from the 
listed directories. For <strong>brukerspectra</strong> data, TREND will read 
<code>1r</code> or <code>2rr</code> files from the <code>pdata1</code> subdirectorires. Therefore, 
the processing number must always be set as <code>1</code>.<br><img src="https://bytebucket.org/chia_hsu/trend/raw/a6b53994c9d44faaff51209d84ba153c3f7805e6/docs/png/png_manual/bruker-dir-index.png?token=c92c125dee2f271e853a31a14eefe9c72ed02195" alt="dir.index.png"><br><code>file.index</code> for Agilent VnmrJ could be used in a simmilar manner, 
except its spectra are saved in the <code>datdir</code> subdirectories.   </li>
<li>[filename] can be the name of a singlematrix data set or a movie file 
for processing.  </li>
</ul>
</li>
<li><p><code>-s [none/noscaling/auto/pareto/]</code></p>
<p>  <code>-s</code> specifies scaling method applied on rows of the data matrix.</p>
<ul>
<li><strong>none</strong> means skip centering and scaling of the rows of the data 
matrix.  </li>
<li><strong>noscaling</strong> means do data centering without scaling. It should 
be accetable in most conditions.  </li>
<li><strong>auto</strong> means Autoscaling. It treats all peaks or features as equally
important but inflates the measurement errors. It is recommended for NMR
spectra in fast and slow exchange regimes.  </li>
<li><strong>pareto</strong> means Pareto scaling. It reduces the importance of large 
peaks but enhances the low, broader peaks. It is recommended for NMR 
spectra in intermediate exchange.  </li>
<li>Definitions of the scaling methods are given in: <em>J. BMC Genomics 2006, 
7, 142<em>. Other scaling options mentioned in this paper, such as __</em></em>vast__
scaling, <strong>range</strong> scaling, <strong>level</strong> scaling are also provided.  </li>
</ul>
</li>
<li><p><code>-r [auto/1e5/3T/5t]</code></p>
<p><code>-r</code> sets the threshold for filtering low intensity regions. The threshold 
is meaningful when processing  2D NMR spectra (ucsf, ft2), but not for FIDs
or images in png format. There are three ways to set the threshold: auto, 
absolute, and manual (set the number of times the noise level).</p>
<ul>
<li><strong>auto</strong> In <strong>auto</strong> mode the program determines noise from the first 
spectrum and set threshold as 4 times of the noise level for autoscaling
and 0.5 times the noise level for Pareto scaling.  </li>
<li><strong>1e5</strong>: When the arguments are set as integral numbers or floating point
numbers , such as <code>1e5</code> or <code>100000</code>. This is used to specify the 
threshold value using the numbering scale of the measurement.  </li>
<li><strong>3T</strong> or <strong>5t</strong>: Numbers with suffix <code>T</code> or <code>t</code> annotates  times of 
noise level of the first spectrum. In this mode threshold is set as 3 or
5 times of noise level of the first spectrum. Three to seven-fold (3T to 
7T) is the recommended range for NMR spectra.  </li>
</ul>
</li>
<li><p><code>-o [output prefix name]</code></p>
<p><code>-o</code> specifies the prefix of output files <code>prefix-s.txt</code>, <code>prefix-vt.txt</code>, 
<code>prefix-pc.txt</code>, and <code>prefix-PC1.txt</code> from PCA. <code>prefix-pc.txt</code> reports all
principle components as successive vertical columns. <code>prefix-PC1.txt</code> 
reports the first principal component as a single vertical column, which in
a titration is the binding isotherm.</p>
<p>For example, the option <code>-o test</code> generates output files named as 
<code>test-s.txt</code>, <code>test-vt.txt</code>, <code>test-PC.txt</code>, <code>test-PC1.txt</code></p>
<p><code>prefix-s.txt</code> and <code>prefix-vt.txt</code> save S and VT matrices in the SVD 
calculation. U matrix is by default omitted because its usually big size.</p>
<p>When ICA is used instead of PCA (<code>-i</code> option is on), the prefix includes 
<code>prefix-vt.txt</code>, which is unmixing matrix, and <code>prefix-IC.txt</code>, which reports
normalized independent components as successive vertical columns.</p>
</li>
<li><p><code>--csv</code> 
When <code>--csv</code> is turned on, data in <code>prefix-s.txt</code>, <code>prefix-pc.txt</code>, 
<code>prefix-PC1.txt</code>, or <code>prefix-IC.txt</code> are generated as CSV files, which can 
be read by software such as Excel or OpenOffice. The corresponding file 
names are <code>prefix-s.csv</code>, <code>prefix-pc.csv</code>, <code>prefix-PC1.csv</code>, <code>prefix-IC.csv</code>.
In csv files (except <code>prefix-s.csv</code>), data is organized as sucessive 
vertical columns, where the first column represents component number. A 
table header is added, such as <code># Component, PC1, PC2, PC3 ...</code>.  </p>
</li>
<li><code>-b</code><br>Traditional uniform binning is applied for NMR data when <code>-b</code> option is 
turned on, it does not improve accuracy but may be useful when the input 
dataset is too large. By default <code>-b</code> option is <code>off</code>. <code>bintimes</code> option 
needs to be set in this condition.  </li>
<li><code>--bintimes [binning time]</code><br>Number of points to be integrated and merged together into a single point. 
The default is 8. For example, if the size of a 2D spectra is 1024*2048, 
setting <code>--bintimes 8</code> causes size of binned spectra to become 
1024/8 <em> 2048/8 = 128 </em> 256  </li>
<li><code>--compress [compress factor]</code><br>When processing <strong>movie</strong> the video can be resized by [compress factor] to 
reduce computational cost. For example, 0.8 means the video size
will be resized to 80%. [compress factor] can be cho  sen from the interval 
(0, 1.0]. By default [compress factor] is set as 1.0, which means use all 
the data, i.e. don&apos;t compress the data.</li>
<li><code>--sparsity [sparsity factor]</code><br>This option controls skipping frames of the input video file by picking up 
every n-th frame from the video, where n is specified by [sparsity factor].
For example, a [sparsity factor] of 1 means all frames will be used. 2 uses 
every second frame, 3 uses every third frame, and so on.  </li>
<li><code>--starttime [start time]</code><br>TREND supports making a subclip of input movie by setting start and end time. 
The default value for <code>--starttime</code> is 0.0. The number format for setting 
start and end time could either be floating point numbers of seconds 
(e.g. <code>0.2</code> stands for 0.2 s), or <code>hh:mm:ss.ff</code>, such as <code>00:03:05.00</code>, 
which means 3 minutes and 5 seconds.  </li>
<li><code>--endtime [end time]</code><br>This option sets the end time for the subclip of input movie. By default 
it is set as <code>end</code>. When <code>--starttime</code> and <code>--endtime</code> are set as default, 
(<code>0.0</code> and <code>end</code>, respectively), the whole input video will be analyzed. 
Otherwise a subclip video will be analyzed and exported as <code>from-starttime-to-endtime_movieclip.mp4</code>, 
or <code>from-starttime_movieclip.mp4</code> if the <code>--endtime</code> is set as the default <code>end</code>.  </li>
<li><code>--columnscaling [none/noscaling/auto/pareto]</code><br><code>--columnscaling</code> specifies the scaling method applied on columns of the 
data matrix. Options and meanings of scaling methods are defined in the <code>-s</code>
option. By default <code>--columnscaling</code> is set as [none].  </li>
<li><code>-i</code> or <code>--ica</code><br>This option uses the independent component analysis (ICA) module instead of
PCA.  </li>
<li><code>--nica</code><br>When ICA is used, the number (n) of components to be calculated must be set
as <code>--nica n</code>. The default n is 2.  </li>
<li><code>--reconst</code><br><code>--reconst</code> is an option for reconstruction. It has no effect on the 
calculation, but dumps necessary files for reconstruction in the future  </li>
<li><code>--keepfiltered</code> is an option for 2D NMR spectra reconstruction when 
<code>--reconst</code> is on. When the input files are 2D NMR data such as ucsf or ft2,
a threshold can be set using the <code>-r</code> option. All data points below the 
threshold are filtered out. <code>--keepfiltered</code> dumps the filtered data points
to a file for reconstruction purpose. This file is not required because 
reconstruction of PCA can be done without the filtered data points.  </li>
<li><code>-x [xaxisfile]</code><br><code>-x</code> specifies the name of file containing the list of tick marks to be used
for the x-axis. It is optional. By default this option is off and x-axis 
ticks are integers. When <code>xaxisfile</code> is specified, numbers within it will 
form the x-axis. The number of rows in <code>xaxisfile</code> should identical to 
<code>index</code> file specified by the  <code>-f [fileindex]</code> option of <code>trendmain.exe</code><br>For example, <code>concentration.txt</code> lists the five concentrations of a 
titration, one per line:<br><img src="https://bitbucket.org/repo/bboE8M/images/1883223853-concentration.txt.png" alt="concentration.txt"></li>
<li><code>-u [Unit]</code><br><code>-u</code> specifies label placed on the X-axis for plotting. [Unit] can be any 
string.  </li>
<li><code>--report</code><br><code>--report</code> generates an HTML file named as <code>prefix-report.html</code> that reports
the results. The report lists arguments used for calculation and files 
created. It plots the first 3 PCs or ICs without normalization. It 
also plots scree plot as well as autocorrelation coefficences that 
represent the smoothness of component curves. For more control of 
plotting of principal or independent components display, please use
<code>trendplot.exe</code>.  </li>
</ul>
</li>
</ul>
<h3 id="plotting-script----display-the-principal-components-identified-by-the-main-script">Plotting script -- display the principal components identified by the main script</h3>
<ul>
<li><h4 id="gui-and-cli-programs">GUI and CLI programs</h4>
<p>GUI: <code>trendplotgui.exe</code> in Windows and Linux, <code>trendplotgui.app</code> in OS X</p>
<p>CLI: <code>trendplot.exe</code> in all platforms</p>
</li>
<li><h4 id="synopsis">Synopsis</h4>
<p><code>trendplot.exe [-r] [-f] [prefix] [-x [xaxisfile]] [-u [Unit]] [-n [the first N components]]
[-p] [-e] -m [0/1/2/3/4/5]] --plot</code></p>
</li>
<li><h4 id="description">Description</h4>
<p><code>trendplot.exe</code> plots the output of <code>trendmain.exe</code>. It is normally used to plot 
first principal component (PC1) or n principal components. It can also graph a 
scree plot of the sizes of the PCs. It uses matplotlib.</p>
</li>
<li><h4 id="options">Options</h4>
<ul>
<li><code>-r</code><br><code>-r</code> reads all arguments used by <code>trendmain.exe</code>. Therefore when <code>-r</code> is used
all <code>trendplot.exe</code> options related to <code>trendmain.exe</code> program, such as <code>-f</code> 
and <code>--ica</code> options can be omitted.  </li>
<li><code>-f [input prefix name]</code><br><code>-f</code> specifies prefix of input files. This needs to be the same as specified 
with the <code>-o</code> option when using the <code>trendmain.exe</code> program. e.g. uM or time (min)</li>
<li><code>-x [xaxisfile]</code><br>The usage is identical to the <code>-x</code> option of <code>trendmain.exe</code>.  </li>
<li><code>-u [Unit]</code><br>The usage is identical to the <code>-u</code> option of <code>trendmain.exe</code>.  </li>
<li><code>-n [integral number N]</code><br><code>-n</code> specifies the number of principal components to plot. For example, 
<code>-n 3</code> plots the first 3 principal components.  </li>
<li><code>--single</code><br><code>--single</code> plots the component number <code>N</code> specified by <code>-n</code> option. For 
example, <code>-n 3 --single</code> plots the third principal component only.  </li>
<li><code>-p</code><br><code>-p</code> plots PC1 values with normalization. This is normally the most 
important trend of change in an experiment. It is used without argument.  </li>
<li><code>-e</code><br><code>-e</code> plots the scree plot, i.e. the size of each of the principal components   </li>
<li><p><code>-m [0/1/2/3/4/5]</code></p>
<p><code>-m</code>  turns on normalization of the <code>N</code> principle components plotted using 
the <code>-n</code> option.  There are several modes of normalization to choose among;
see the table below for the choices. When <code>-m</code> option is turned on, the 
normalized PC values will be saved to <code>normalizedPC.csv</code> file.  </p>
</li>
</ul>
</li>
</ul>
<table>
<thead>
<tr>
<th>Normalization mode</th>
<th>meaning</th>
</tr>
</thead>
<tbody>
<tr>
<td>0</td>
<td>No normalization on PC values</td>
</tr>
<tr>
<td>1</td>
<td>Normalize all PC values to 1.0, with PC1 increasing (typical for a ligand binding isotherm)</td>
</tr>
<tr>
<td>2</td>
<td>Normalize all PC values to 1.0, regardless of the direction</td>
</tr>
<tr>
<td>3</td>
<td>Normalize PC1 to 1.0 and set it to increase. Normalize all other components to PC1</td>
</tr>
<tr>
<td>4</td>
<td>Normalize PC1 to 1.0 without changing its orientation. Normalize all other components to PC1</td>
</tr>
<tr>
<td>5*</td>
<td>Normalize PC values to the sum of PC1+PC2</td>
</tr>
<tr>
<td>6<sup>$</sup></td>
<td>Normalize 2 phase binding isotherms into monotonic increasing, reverse turning point</td>
</tr>
</tbody>
</table>
<p>* This mode was used for two sequential binding events
$  This option might be used for binding isotherms of titrations where 
    spectra in the middle of the titration are severely perturbed and the 
    trajectory of change is V-shaped.</p>
<ul>
<li><p><code>--biplot</code></p>
<p>This option shows the &quot;bi-plot&quot; of the first two principal components in 
matrix VT. (A bi-plot is the equivalent of a score plot in the nonmenclature
of the metabolomics field).</p>
</li>
<li><p><code>--ica</code></p>
<p>This option is required for plotting ICA results. All <em>principal components</em>
mentioned above will be replaced by <em>independent components</em> when <code>--ica</code> is
turned on.</p>
</li>
<li><p><code>--plot</code><br><code>--plot</code> is optional. When it is on, all figures will be saved as PNG images.  </p>
</li>
</ul>
<h3 id="reconstruction-script---use-the-pca-results-from-the-main-script-to-model-the-2d-measurements">Reconstruction script - Use the PCA results from the Main script to model the 2D measurements</h3>
<ul>
<li><h4 id="gui-and-cli-programs">GUI and CLI programs</h4>
<p>GUI: <code>trendreconstructgui.exe</code> in Windows and Linux, <code>trendreconstructgui.app</code>
in OS X</p>
<p>CLI: <code>trendreconstruct.exe</code> in all platforms</p>
</li>
<li><h4 id="synopsis">Synopsis</h4>
<pre><code>trendreconstruct.exe -n [components] -m [1/2/3/4]] [-e] [-p] [-r] [-z] 
[-d] [--rmsd] [--gif] [--dpi integral for dpi of output PNG files]
</code></pre></li>
<li><h4 id="description">Description</h4>
<p><code>trendreconstruct.exe</code> rebuilds the original 2D measurements using the first
   N principal components or the Nth principal component obtained from the 
   main script. The script <code>trendreconstruct.exe</code> requires that the 
   <code>--reconstruct</code> option be turned on when using <code>trendmain.exe</code>. Normally 
   <code>trendmain.exe</code> will not save the files needed for PCA reconstruction unless
   the  <code>--reconst</code> option is used.</p>
</li>
<li><h4 id="options">Options</h4>
<ul>
<li><code>-n component(s) selected for reconstruction</code><br><code>-n</code> specifies principal components selected to use in rebuilding 
the original measurements. For example, <code>-n 1</code> means reconstruct the 2D 
measurements using only the first component, while <code>-n 3</code> uses the 
third components. To reconstruct individual, non-consecutive components, 
you can separate the page numbers with commas: e.g. <code>-n 2,4,6</code> would 
reconstruct using the second, fourth, and sixth components. To reconstruct 
the range of components 1 to 3, use <code>-n 1-3</code>. Range and individual 
components can be selected together, e.g. <code>-n 1,3-5,7</code> means reconstruction 
by components 1, 3, 4, 5, 7. Note space is <strong>not</strong> allowed between selected 
components unless quotation marks are used. i.e. <code>-n &apos;1, 3-5, 7&apos;</code>.   </li>
<li><p><code>-m [1/2/3/4]</code></p>
<p>Background:  <code>trendmain.exe</code> reorganizes the measurements listed in 
<code>file.index</code> as a large 2D matrix, filters them, and scales them for PCA. 
The <code>-m</code> option specifies which stage of the process should be reconstructed. 
<code>-m</code> specifies which matrix will be reconstructed. The choices are:  </p>
</li>
</ul>
</li>
</ul>
<table>
<thead>
<tr>
<th>matrix mode</th>
<th>meaning</th>
</tr>
</thead>
<tbody>
<tr>
<td>1</td>
<td>This recovers the state right before doing PCA. The large 2D matrix is row (column) scaled, with unchanged rows and noisy regions filtered out</td>
</tr>
<tr>
<td>2</td>
<td>This differs from matrix mode 1 in that the rows (columns) are unscaled</td>
</tr>
<tr>
<td>3</td>
<td>This option also recovers the filtered, unchanged rows from matrix mode 2</td>
</tr>
<tr>
<td>4*</td>
<td>This option differs from matrix mode 3 in also recovering the low intensity regions filtered out from being below the threshold</td>
</tr>
</tbody>
</table>
<p>* mode 4 is for <strong>Fourier-Transformed 2D NMR spectra</strong> where <code>-r</code> option 
    was used in <code>trendmain.exe</code> to set the threshold for filtering low intensity
    regions.</p>
<ul>
<li><p><code>-e</code></p>
<p><code>-e</code> is optional. When it is on, the reconstructions of the measurements are
exported in the same format as the original measurements read into 
<code>trendmain.exe</code>. If input format is <code>movie</code> or <code>png</code>, the output 
format will be mp4 format movie clip. mp4 movie reconstructed from <code>movie</code> 
shares the same time length as the input movie. mp4 movie reconstructed 
from a series of PNG images is set as 10 second long.</p>
</li>
<li><p><code>-p</code></p>
<p><code>-p</code> is optional. When it is on, the reconstructed images are exported as 
images in PNG format.</p>
</li>
<li><p><code>--gif</code><br><code>--gif</code> is optional. When it is on, the reconstructed movies are 
exported as GIF animation just in case there is no appropriate 
media player to play <code>.mp4</code> movie clip generated by <code>-e</code>. GIF animation 
is more compatible since most web browsers can play it, although it 
has larger size than mp4.  </p>
</li>
<li><code>--dpi</code><br>When <code>-p</code> option is turned on, <code>--dpi</code> specifies the dots-per-inch resolution
of the png files to be reconstructed. By default, this is set to 96. This 
option is not required.  </li>
<li><code>-z</code><br><code>-z</code> option is used to recover rows that do not change and hence was filtered
out before doing PCA.  </li>
<li><code>-r</code><br><code>-r</code> is used only when reconstructing 2D Fourier-transformed NMR spectra. 
When doing PCA on 2D NMR spectra, a threshold can be set to filter out noisy
regions using the <code>-r</code> option for <code>trendmain.exe</code>. The <code>-r</code> option can be 
used in reconstruction to recover those noisy regions in order to fully 
recover the original spectra. It is not required and is usually not necessary.  </li>
<li><code>-d</code> or <code>--diff</code><br><code>-d</code> turns on the <code>diff</code> mode, exporting difference spectra instead of 
reconstructed spectra. It is used to visualize the difference between original
and reconstructed spectra.  </li>
<li><code>--rmsd</code><br><code>--rmsd</code> calculates the RMSD between the reconstructed and original data for
all principal components. A file with suffix name as <code>-rmsd.txt</code> and prefix
name specified by the <code>-o</code> option of <code>trendmain.exe</code> will be stored. For 
example, if <code>trendmain.exe</code> was used with the <code>-o example</code> option, <code>--rmsd</code> 
option generates a file named as <code>prefix-rmsd.txt</code>.  </li>
<li><code>-o [output prefix name]</code><br><code>-o</code> specifies the prefix of reconstructed files or images  </li>
</ul>
<h3 id="svd-test-script">SVD test script</h3>
<ul>
<li><h4 id="cli-program">CLI program</h4>
<p><code>trendtest.exe</code> in all platforms</p>
</li>
<li><h4 id="synopsis">Synopsis</h4>
<pre><code>trendtest.exe [-r] [-f [outname prefix of trendmain.exe]]
</code></pre></li>
<li><h4 id="description">Description</h4>
<p><code>trendtest.exe</code> analyzes PCA results generated by <code>trendmain.exe</code> and <code>reconst.py</code>.
It saves a table of attributes for evaluating the SVD outcome in the 
<code>prefix-svdtest.csv</code> file. Columns of the table are: Component, Size of singular
value, cumulative contribution ratio, auto correlation coefficient, RMSD, and 
Normalized RMSD, respectively. The last two columns are optional and depend on 
whether <code>trendreconstruct.exe --rmsd</code> was executed.</p>
</li>
<li><h4 id="options">Options</h4>
<ul>
<li><code>-r</code><br><code>-r</code> reads all arguments used by <code>trendmain.exe</code>. When <code>-r</code> is used <code>-f</code> is 
not necessary  </li>
<li><code>-f [input prefix name]</code><br><code>-f</code> specifies the prefix of input files. This needs to be the same as specified 
with the <code>-o</code> option when using the <code>trendmain.exe</code> program  </li>
</ul>
</li>
</ul>
<h3 id="report-script">Report script</h3>
<ul>
<li><h3 id="cli-program">CLI program</h3>
<p><code>trendplot.exe</code> in all platforms.</p>
</li>
<li><h4 id="synopsis">Synopsis</h4>
<p><code>trendreport.exe</code></p>
</li>
<li><h4 id="description">Description</h4>
<p><code>trendreport.exe</code> generates an HTML report just in case the <code>trendmain</code> 
was executed without turning on <code>--report</code> option.</p>
</li>
</ul>

                                
                                </section>
                            
    </div>
    <div class="search-results">
        <div class="has-results">
            
            <h1 class="search-results-title"><span class='search-results-count'></span> results matching "<span class='search-query'></span>"</h1>
            <ul class="search-results-list"></ul>
            
        </div>
        <div class="no-results">
            
            <h1 class="search-results-title">No results matching "<span class='search-query'></span>"</h1>
            
        </div>
    </div>
</div>

                        </div>
                    </div>
                
            </div>

            
                
                <a href="GUI.html" class="navigation navigation-prev " aria-label="Previous page: GUI manual">
                    <i class="fa fa-angle-left"></i>
                </a>
                
                
                <a href="tutorial.html" class="navigation navigation-next " aria-label="Next page: Tutorial">
                    <i class="fa fa-angle-right"></i>
                </a>
                
            
        
    </div>

    <script>
        var gitbook = gitbook || [];
        gitbook.push(function() {
            gitbook.page.hasChanged({"page":{"title":"Command Line Manual","level":"1.5","depth":1,"next":{"title":"Tutorial","level":"1.6","depth":1,"path":"doc/tutorial.md","ref":"doc/tutorial.md","articles":[]},"previous":{"title":"GUI manual","level":"1.4","depth":1,"path":"doc/GUI.md","ref":"doc/GUI.md","articles":[]},"dir":"ltr"},"config":{"gitbook":"*","theme":"default","variables":{},"plugins":["livereload"],"pluginsConfig":{"livereload":{},"highlight":{},"search":{},"lunr":{"maxIndexSize":1000000,"ignoreSpecialCharacters":false},"sharing":{"facebook":true,"twitter":true,"google":false,"weibo":false,"instapaper":false,"vk":false,"all":["facebook","google","twitter","weibo","instapaper"]},"fontsettings":{"theme":"white","family":"sans","size":2},"theme-default":{"styles":{"website":"styles/website.css","pdf":"styles/pdf.css","epub":"styles/epub.css","mobi":"styles/mobi.css","ebook":"styles/ebook.css","print":"styles/print.css"},"showLevel":false}},"structure":{"langs":"LANGS.md","readme":"README.md","glossary":"GLOSSARY.md","summary":"SUMMARY.md"},"pdf":{"pageNumbers":true,"fontSize":12,"fontFamily":"Arial","paperSize":"a4","chapterMark":"pagebreak","pageBreaksBefore":"/","margin":{"right":62,"left":62,"top":56,"bottom":56}},"styles":{"website":"styles/website.css","pdf":"styles/pdf.css","epub":"styles/epub.css","mobi":"styles/mobi.css","ebook":"styles/ebook.css","print":"styles/print.css"}},"file":{"path":"doc/manual.md","mtime":"2016-11-08T23:03:02.000Z","type":"markdown"},"gitbook":{"version":"3.2.2","time":"2016-11-08T22:58:36.051Z"},"basePath":"..","book":{"language":""}});
        });
    </script>
</div>

        
    <script src="../gitbook/gitbook.js"></script>
    <script src="../gitbook/theme.js"></script>
    
        
        <script src="../gitbook/gitbook-plugin-livereload/plugin.js"></script>
        
    
        
        <script src="../gitbook/gitbook-plugin-search/search-engine.js"></script>
        
    
        
        <script src="../gitbook/gitbook-plugin-search/search.js"></script>
        
    
        
        <script src="../gitbook/gitbook-plugin-lunr/lunr.min.js"></script>
        
    
        
        <script src="../gitbook/gitbook-plugin-lunr/search-lunr.js"></script>
        
    
        
        <script src="../gitbook/gitbook-plugin-sharing/buttons.js"></script>
        
    
        
        <script src="../gitbook/gitbook-plugin-fontsettings/fontsettings.js"></script>
        
    

    </body>
</html>

